File: sound-pool/src/main/java/com/olekdia/soundpool/SoundSampleDescriptor.java
package com.olekdia.soundpool;

import android.content.ContentResolver;
import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.net.Uri;
import android.os.ParcelFileDescriptor;

import com.olekdia.androidcommon.extensions.FileExtensionsKt;

import java.io.Closeable;
import java.io.File;
import java.io.FileDescriptor;
import java.io.FileNotFoundException;
import java.io.IOException;

public class SoundSampleDescriptor implements Closeable {

    public final long mFileOffset;
    public final long mFileSize;

    private AssetFileDescriptor mAssetDescr;
    private ParcelFileDescriptor mParcelDescr;
    private FileDescriptor mFd;

    public SoundSampleDescriptor(
        final Context context,
        final SoundSampleMetadata metadata
    ) throws FileNotFoundException {
        if (metadata.getPath() == null) {
            mParcelDescr = null;

            mAssetDescr = context.getResources().openRawResourceFd(metadata.getRawResId());
            mFileOffset = mAssetDescr.getStartOffset();
            mFileSize = mAssetDescr.getLength();
            mFd = mAssetDescr.getFileDescriptor();
        } else if (FileExtensionsKt.isFileDocUri(metadata.getPath())) {
            final ContentResolver cr = context.getContentResolver();
            final Uri uri = Uri.parse(metadata.getPath());
            mFileOffset = 0;
            mFileSize = FileExtensionsKt.getFileSize(uri, cr);
            mParcelDescr = cr.openFileDescriptor(uri, "r");
            mFd = mParcelDescr.getFileDescriptor();
        } else {
            mAssetDescr = null;

            final File file = new File(metadata.getPath());
            mParcelDescr = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY);
            mFileOffset = 0;
            mFileSize = file.length();
            mFd = mParcelDescr.getFileDescriptor();
        }
    }

    public final FileDescriptor getFileDescriptor() {
        return mFd;
    }

    @Override
    public void close() throws IOException {
        if (mParcelDescr != null) {
            mParcelDescr.close();
        } else if (mAssetDescr != null) {
            if (mAssetDescr instanceof Closeable) {
                mAssetDescr.close();
            } else if (mAssetDescr.getParcelFileDescriptor() instanceof Closeable) {
                mAssetDescr.getParcelFileDescriptor().close();
            }
        }
        mAssetDescr = null;
        mParcelDescr = null;
        mFd = null;
    }
}

File: sound-pool/src/main/java/com/olekdia/soundpool/SoundSampleDescriptor.kt
package com.olekdia.soundpool

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.media.MediaExtractor
import android.net.Uri
import android.os.ParcelFileDescriptor
import com.olekdia.androidcommon.NO_RESOURCE
import com.olekdia.androidcommon.extensions.getFileSize
import com.olekdia.common.INVALID_L
import java.io.Closeable
import java.io.File
import java.io.FileDescriptor
import java.io.IOException
import java.net.URL

class SoundSampleDescriptor @Throws(IOException::class)
constructor(
    context: Context,
    private val metadata: SoundSampleMetadata
) : Closeable {

    /**
     * File size in bytes
     */
    val fileSize: Long
    private val fileOffset: Long

    private var assetDescr: AssetFileDescriptor? = null
    private var parcelDescr: ParcelFileDescriptor? = null

    private var httpPath: String? = null
    private var fileDescriptor: FileDescriptor? = null

    init {
        when {
            metadata.rawResId != NO_RESOURCE -> {
                parcelDescr = null
                assetDescr = context.resources.openRawResourceFd(metadata.rawResId)
                    .also {
                        fileOffset = it.startOffset
                        fileSize = it.length
                        fileDescriptor = it.fileDescriptor
                    }
            }

            metadata.path.isContentPath -> {
                val cr = context.contentResolver
                val uri = Uri.parse(metadata.path.fullPath)

                parcelDescr = cr.openFileDescriptor(uri, "r")
                    .also {
                        fileOffset = 0
                        fileSize = uri.getFileSize(cr)
                        fileDescriptor = it!!.fileDescriptor
                    }
            }

            metadata.path.isHttpPath -> {
                httpPath = metadata.path.fullPath
                fileOffset = 0
                fileSize = URL(httpPath).getFileSize().let {
                    if (it == INVALID_L) {
                        DEFAULT_BUFFER_SIZE.toLong()
                    } else {
                        it
                    }
                }
            }

            metadata.path.isFilePath -> {
                assetDescr = null

                val file = File(metadata.path.fullPath)
                parcelDescr = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
                    .also {
                        fileOffset = 0
                        fileSize = file.length()
                        fileDescriptor = it.fileDescriptor
                    }
            }

            else -> {
                fileOffset = 0
                fileSize = 0
            }
        }
    }

    fun setExtractorDataSource(extractor: MediaExtractor) {
        fileDescriptor?.let {
            extractor.setDataSource(it, fileOffset, fileSize)
        } ?: httpPath?.let {
            extractor.setDataSource(it)
        } ?: run {
            throw IOException()
        }
    }

    @Throws(IOException::class)
    override fun close() {
        parcelDescr?.close()
        assetDescr?.let {
            if (it is Closeable) {
                it.close()
            } else if (it.parcelFileDescriptor is Closeable) {
                it.parcelFileDescriptor.close()
            }
        }
        assetDescr = null
        parcelDescr = null
    }
}